src/network/core/address.cpp

```diff
@@ -317,7 +317,7 @@ static SOCKET ConnectLoopProc(addrinfo *runp)
 
 	SOCKET sock = socket(runp->ai_family, runp->ai_socktype, runp->ai_protocol);
 	if (sock == INVALID_SOCKET) {
-		DEBUG(net, 1, "[%s] could not create %s socket: %s", type, family, NetworkGetLastErrorString());
+		DEBUG(net, 1, "[%s] could not create %s socket: %s", type, family, NetworkError::GetLastAsString().c_str());
```

Comment:
I'm not sure this works, you are calling `NetworkError::GetLastAsString()` which creates a temporary `std::string`, then calling `.c_str()` to get the pointer to the allocated string without storing the `std::string` in a variable. I'm not entirely sure about this, but I think the compiler is allowed to de-allocate the string before it's passed into your logging function (since it's a temporary, isn't stored anywhere, and the compiler has no way of knowing that running the sting destructor will make the `char *` you got point to invalid memory).

Reply:
The specification says: `Temporary objects are destroyed as the last step in evaluating the full-expression (6.8.1) that (lexically) contains the point where they were created` (15.2.4). So this code is fine. Assigning the `const char *` of `c_str()` to a variable and then passing it to DEBUG is definitely not fine.
See also https://en.cppreference.com/w/cpp/language/lifetime and look for "Temporary object", as that's what the std::string getting returned here is. What might be easier to read would be https://lesleylai.info/en/temporaries/

Reply:
What I was suggesting was assigning the std::string itself to a variable, then passing in `str.c_str()` to DEBUG while the std::string was in scope.

TIL about the guarantees of temporary variables, I dodn't know about that, thanks for the tip!